### PR TITLE
[DFT][MKLGPU] Update mklgpu support to mkl preparing for oneapi 2025.0 release

### PR DESCRIPTION
**Pull request by s-Nick**
# Description

Update how dft mklgpu backend includes header files from MKL and how cmake adds them to compilation. 
This is done to avoid future name conflicts between this library and backend library. Forcing the compiler to look for some headers following the "" and <> inclusions rules accordingly with [C++ Core Guideline SF.12](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-incform)
# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
[arc_mklgpu_dft_log.txt](https://github.com/user-attachments/files/17161322/arc_mklgpu_dft_log.txt)
**N.B.** Tests ran with current oneAPI release (2024.2) I cannot run tests with next one.

- [x] Have you formatted the code using clang-format?

### CODE REVIEW

File: src/dft/backends/mklgpu/CMakeLists.txt [hunk 1]
```diff
@@ -32,12 +32,14 @@ add_library(${LIB_OBJ} OBJECT
 )
 add_dependencies(onemkl_backend_libs_dft ${LIB_NAME})
 
-target_include_directories(${LIB_OBJ}
-  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
-)
```
Comment (s-Nick) [inline src/dft/backends/mklgpu/CMakeLists.txt:37]: Giving the includes change I made a little bit down. This command is not necessary anymore, so I decided to remove it trying to keep the cmake as lean as possible.

File: src/dft/backends/mklgpu/CMakeLists.txt [hunk 2]
```diff
@@ -32,12 +32,14 @@ add_library(${LIB_OBJ} OBJECT
...
 target_include_directories(${LIB_NAME}
   PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
 )
+
+target_compile_options(${LIB_OBJ}
+  BEFORE PRIVATE -iquote $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
```
Comment (Rbiessy) [inline src/dft/backends/mklgpu/CMakeLists.txt:44]: This deserves a comment to explain the header conflict and why the `iquote` is needed.
That looks better than I expected, this is great!

**Review by Rbiessy — COMMENTED** [review]
LGTM, thanks for this.
I'll approve once we have been able to test this with nightly builds.

File: src/dft/backends/mklgpu/CMakeLists.txt (hunk 2 above)
Comment (s-Nick) [reply src/dft/backends/mklgpu/CMakeLists.txt:44]: Added comment in [562f1db](https://github.com/oneapi-src/oneMKL/pull/575/commits/562f1dbdc763fd65107eb3247b30fdfc053189b0)

**Review by Rbiessy — APPROVED** [review]

**Comment by s-Nick** [comment]
Log using next MKL 
[dft_mklgpu_2025_log.txt](https://github.com/user-attachments/files/17296156/dft_mklgpu_2025_log.txt)

File: src/dft/backends/mklgpu/CMakeLists.txt [hunk 3]
```diff
@@ -32,12 +32,17 @@ add_library(${LIB_OBJ} OBJECT
...
 
-target_include_directories(${LIB_OBJ}
-  PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
-)
 target_include_directories(${LIB_NAME}
   PUBLIC ${ONEMKL_INTERFACE_INCLUDE_DIRS}
 )
+
+# Due to using the same file name for different headers in this library and in
+# MKL Intel library, we force the compiler to follow C++ Core Guideline SF.12
```
Comment (lhuot) [inline src/dft/backends/mklgpu/CMakeLists.txt]:
```suggestion
# the Intel(R) oneAPI Math Kernel Library, we force the compiler to follow C++ Core Guideline SF.12
```

**Review by lhuot — APPROVED** [review]
LGTM, minor comment added. Thank you!

**Review by Rbiessy — APPROVED** [review]

File: src/dft/backends/mklgpu/CMakeLists.txt (hunk 3 above)
Comment (s-Nick) [reply src/dft/backends/mklgpu/CMakeLists.txt]: Thank you @lhuot I updated the comment in [18bd3da](https://github.com/oneapi-src/oneMKL/pull/575/commits/18bd3da50a093f7826408b1c19384749ae93b52c)

File: tests/unit_tests/dft/source/compute_tests.cpp [hunk 4]
```diff
@@ -75,7 +75,8 @@ class ComputeTests_real_real_out_of_place_REAL
         }                                                                                 \
         catch (std::exception & e) {                                                      \
             std::string msg = e.what();                                                   \
-            if (msg.find("FFT_UNIMPLEMENTED") != std::string::npos) {                     \
+            if ((msg.find("FFT_UNIMPLEMENTED") != std::string::npos) ||                   \
+                (msg.find("Unimplemented") != std::string::npos)) {                       \
```
Comment (raphael-egan) [inline tests/unit_tests/dft/source/compute_tests.cpp:79]: @s-Nick, was this addition motivated by a new exception that's _not_ a `oneapi::mkl::unimplemented` exception with "Unimplemented" in its message? If yes, could you please share details?

**Pull request merged**